win32ss/user/ntuser/window.c

```diff
@@ -3352,98 +3352,79 @@ NtUserFindWindowEx(HWND hwndParent,
 }
 
 
-/*
- * @implemented
- */
+/* @implemented */
 PWND FASTCALL UserGetAncestor(PWND Wnd, UINT Type)
 {
-   PWND WndAncestor, Parent;
+    PWND Ret, Parent, pwndNode, pwndMessage;
+    PDESKTOP pDesktop = Wnd->head.rpdesk;
+    PWND pwndDesktop = pDesktop->pDeskInfo->spwnd;
 
-   if (UserHMGetHandle(Wnd) == IntGetDesktopWindow())
-   {
-      return NULL;
-   }
+    if (Wnd == pwndDesktop)
+        return NULL;
 
-   switch (Type)
-   {
-      case GA_PARENT:
-         {
-            WndAncestor = Wnd->spwndParent;
-            break;
-         }
+    pwndMessage = pDesktop->spwndMessage;
+    if (Wnd == pwndMessage)
+        return NULL;
 
-      case GA_ROOT:
-         {
-            WndAncestor = Wnd;
-            Parent = NULL;
+    Parent = Wnd->spwndParent;
+    if (!Parent)
+        return NULL;
 
-            for(;;)
+    switch (Type)
+    {
+        case GA_PARENT:
+            return Parent;
+        case GA_ROOT:
+            Ret = Wnd;
+            if (Parent == pwndDesktop)
+                break;
+            do
             {
-               if(!(Parent = WndAncestor->spwndParent))
-               {
-                  break;
-               }
-               if(IntIsDesktopWindow(Parent))
-               {
-                  break;
-               }
-
-               WndAncestor = Parent;
-            }
+                if (Parent == pwndMessage)
+                    break;
+                Ret = Parent;
+                pDesktop = Parent->head.rpdesk;
+                Parent = Parent->spwndParent;
+            } while (Parent != pDesktop->pDeskInfo->spwnd);
```

Comment:
The "current" pDesktop can change there?

Reply:
Yes, it can. The desktop can be multiple.